src/Raven.Server/Documents/CountersStorage.cs

```diff
@@ -1820,6 +1907,253 @@ public string DeleteCounter(DocumentsOperationContext context, string documentId
+            foreach ((_, Table.TableValueHolder tvh) in table.SeekByPrimaryKeyPrefix(documentKeyPrefix, Slices.Empty, 0))
+            {
+                var item = TableValueToCounterTombstoneDetail(context, ref tvh.Reader);
+                yield return item;
+            }
+        }
+
+        public IEnumerable<CounterTombstoneDetail> GetCounterTombstonesFrom(DocumentsOperationContext context, long etag, long toEtag = long.MaxValue)
+        {
+            var table = new Table(CounterTombstonesSchema, context.Transaction.InnerTransaction);
+
+            // ReSharper disable once LoopCanBeConvertedToQuery
+            foreach (var result in table.SeekForwardFrom(CounterTombstonesSchema.FixedSizeIndexes[AllCounterTombstonesEtagSlice], etag, 0))
+            {
+                var item = TableValueToCounterTombstoneDetail(context, ref result.Reader);
+                if (item.Etag > toEtag)
+                    yield break;
+                yield return item;
+            }
+        }
+
+        public long PurgeCountersAndCounterTombstones(DocumentsOperationContext context, string collection, long upto, long numberOfEntriesToDelete)
+        {
+            var collectionName = _documentsStorage.GetCollection(collection, throwIfDoesNotExist: false);
+            if (collectionName == null)
+                return 0;
+
+            var deletedCounters = PurgeCounters(upto, context, collectionName, numberOfEntriesToDelete);
+            var deletedTombstoneEntries = PurgeCounterTombstones(upto, context, collectionName, numberOfEntriesToDelete - deletedCounters);
+            return deletedTombstoneEntries + deletedCounters;
+        }
+
+        private long PurgeCounters(long upto, DocumentsOperationContext context, CollectionName collectionName, long numberOfEntriesToDelete)
+        {
+            var counterTombstonesTableName = collectionName.GetTableName(CollectionTableType.CounterTombstones);
+            var tombstonesTable = context.Transaction.InnerTransaction.OpenTable(CounterTombstonesSchema, counterTombstonesTableName);
+            if (tombstonesTable == null || tombstonesTable.NumberOfEntries == 0 || numberOfEntriesToDelete <= 0)
+                return 0;
+
+            var table = GetCountersTable(context.Transaction.InnerTransaction, collectionName);
+            if (table == null || table.NumberOfEntries == 0)
+                return 0;
+
+            var deleted = 0L;
+
+            foreach (var item in GetCounterTombstonesFrom(context, 0, upto))
+            {
+                // first, we extract the document Id and the deleted counter name from the counter tombstones entry
+                // then, we extract the data from the counters table, by the counter group key
+                // and remove the deleted counter (change the data)
+                // at the end, we'll write the new data to disk and update the document
+
+                if (item.Etag > upto)
```

Comment:
do you take into account `numberOfEntriesToDelete` here?

Reply:
done